File: nfc_proof_of_concept/packages/nfc_host_card_emulation/android/src/main/kotlin/io/flutter/plugins/nfc_host_card_emulation/app_layer/file_access/fields/command_fields.kt
package io.flutter.plugins.nfc_host_card_emulation.file_access.fields

import io.flutter.plugins.nfc_host_card_emulation.app_layer.ApduField

class ApduClass private constructor(claByte: Int) : ApduField("CLA", 1) {
    companion object {
        val standard = ApduClass(0x00)
    }

    init {
        buffer[0] = claByte.toByte()
    }
}

class ApduInstruction(insByte: Int, name: String) : ApduField(name, 1) {
    companion object {
        const val SELECT_BYTE = 0xA4
        const val READ_BINARY_BYTE = 0xB0
        const val UPDATE_BINARY_BYTE = 0xD6

        val select = ApduInstruction(SELECT_BYTE, "INS (SELECT)")
        val readBinary = ApduInstruction(READ_BINARY_BYTE, "INS (READ_BINARY)")
        val updateBinary = ApduInstruction(UPDATE_BINARY_BYTE, "INS (UPDATE_BINARY)")
    }

    init {
        buffer[0] = insByte.toByte()
    }
}

class ApduParams(p1: Int, p2: Int, name: String) : ApduField(name, 2) {
    companion object {
        val byName = ApduParams(0x04, 0x00, "P1-P2 (ByName)")
        val byFileId = ApduParams(0x00, 0x0C, "P1-P2 (ByFileID)")

        fun forOffset(offset: Int): ApduParams {
            require(offset in 0..0xFFFF) { "Offset must be a 16-bit unsigned integer." }
            val p1 = offset shr 8 and 0xFF
            val p2 = offset and 0xFF
            return ApduParams(p1, p2, "P1-P2 (Offset)")
        }
    }

    init {
        buffer[0] = p1.toByte()
        buffer[1] = p2.toByte()
    }
}

class ApduLc(lc: Int) : ApduField("Lc", 1) {
    init {
        require(lc in 0..255) { "Lc must be an 8-bit unsigned integer." }
        buffer[0] = lc.toByte()
    }
}

class ApduLe(le: Int) : ApduField("Le", 1) {
    init {
        require(le in 0..255) { "Le must be an 8-bit unsigned integer. Use 0 for 256 bytes." }
        buffer[0] = le.toByte()
    }
}

File: nfc_proof_of_concept/packages/flutter_hce/lib/app_layer/file_access/fields/command_fields.dart
import '../../field.dart';

class ApduClass extends ApduField {
  static final standard = ApduClass._internal(0x00);

  ApduClass._internal(int claByte) : super(size: 1, name: "CLA") {
    buffer[0] = claByte;
  }
}

class ApduInstruction extends ApduField {
  static const int SELECT_BYTE = 0xA4;
  static const int READ_BINARY_BYTE = 0xB0;
  static const int UPDATE_BINARY_BYTE = 0xD6;

  static final select =
      ApduInstruction._internal(SELECT_BYTE, name: "INS (SELECT)");
  static final readBinary =
      ApduInstruction._internal(READ_BINARY_BYTE, name: "INS (READ_BINARY)");
  static final updateBinary = ApduInstruction._internal(UPDATE_BINARY_BYTE,
      name: "INS (UPDATE_BINARY)");

  ApduInstruction._internal(int insByte, {required String name})
      : super(size: 1, name: name) {
    buffer[0] = insByte;
  }

  /// Smart constructor that tries to return existing static final instances first
  factory ApduInstruction(int insByte, {String? name}) {
    // Try to match with existing static final instances
    switch (insByte) {
      case SELECT_BYTE:
        return select;
      case READ_BINARY_BYTE:
        return readBinary;
      case UPDATE_BINARY_BYTE:
        return updateBinary;
      default:
        // If no match found, create new instance
        final effectiveName = name ??
            "INS (0x${insByte.toRadixString(16).padLeft(2, '0').toUpperCase()})";
        return ApduInstruction._internal(insByte, name: effectiveName);
    }
  }
}

class ApduParams extends ApduField {
  static final byName =
      ApduParams._internal(p1: 0x04, p2: 0x00, name: "P1-P2 (ByName)");
  static final byFileId =
      ApduParams._internal(p1: 0x00, p2: 0x0C, name: "P1-P2 (ByFileID)");

  ApduParams._internal({required int p1, required int p2, required String name})
      : super(size: 2, name: name) {
    buffer[0] = p1;
    buffer[1] = p2;
  }

  /// Smart constructor that tries to return existing static final instances first
  factory ApduParams({required int p1, required int p2, String? name}) {
    // Try to match with existing static final instances
    if (p1 == 0x04 && p2 == 0x00) {
      return byName;
    } else if (p1 == 0x00 && p2 == 0x0C) {
      return byFileId;
    }

    // If no match found, create new instance
    final effectiveName = name ??
        "P1-P2 (0x${p1.toRadixString(16).padLeft(2, '0').toUpperCase()}, 0x${p2.toRadixString(16).padLeft(2, '0').toUpperCase()})";
    return ApduParams._internal(p1: p1, p2: p2, name: effectiveName);
  }

  factory ApduParams.forOffset(int offset) {
    if (offset < 0 || offset > 0xFFFF) {
      throw ArgumentError(
          'Offset must be a 16-bit unsigned integer (0-65535).');
    }
    final p1 = (offset >> 8) & 0xFF;
    final p2 = offset & 0xFF;
    return ApduParams._internal(p1: p1, p2: p2, name: "P1-P2 (Offset)");
  }
}

class ApduLc extends ApduField {
  ApduLc({required int lc}) : super(size: 1, name: "Lc") {
    if (lc < 0 || lc > 255) {
      throw ArgumentError('Lc must be an 8-bit unsigned integer (0-255).');
    }
    buffer[0] = lc;
  }
}

class ApduLe extends ApduField {
  ApduLe({required int le}) : super(size: 1, name: "Le") {
    if (le < 0 || le > 255) {
      throw ArgumentError(
          'Le must be an 8-bit unsigned integer (0-255). Use 0 for 256 bytes.');
    }
    buffer[0] = le;
  }
}
